### DISCUSSION TITLE
TestClient Fails to Call Endpoint with OAuth2PasswordRequestForm Using APIRouter

### DISCUSSION THREAD

**Discussion by IlyaMichlin** [fastapi/fastapi#11802] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import FastAPI, APIRouter, Depends
from fastapi.testclient import TestClient
from fastapi.security import OAuth2PasswordRequestForm


app = FastAPI()
router = APIRouter()
app.include_router(router)


# @app.post("/token")  # working
@router.post("/token")  # not working
def login_for_access_token(
    form_data: Annotated[OAuth2PasswordRequestForm, Depends()],
):
    return form_data


client = TestClient(app)


def test_token_success():
    response = client.post(
        "/token",
        data={"username": "johndoe", "password": "secret"}
    )
    print(f"response: {response.json()}")
    assert response.status_code == 200
```

### Description

I encountered an issue when creating an endpoint that includes OAuth2PasswordRequestForm in its arguments and is described using @router.post(). The endpoint call breaks when using TestClient, though the code works fine when running the app with uvicorn.

When attempting to run the test using TestClient, the following error is encountered: `AttributeError: 'ProactorEventLoop' object has no attribute '_compute_internal_coro'
`.

The endpoint functions correctly when the app is executed with uvicorn. The issue only occurs when the endpoint is defined using @router.post(). Changing @router.post() to @app.post() resolves the issue.

Has anyone experienced a similar issue? Any insights or solutions would be greatly appreciated.

### Operating System

Linux, Windows

### FastAPI Version

0.111.0

### Pydantic Version

2.8.2

### Python Version

3.12.4

### Additional Context

fastapi==0.111.0
httpx==0.27.0
pytest

**Comment by YuriiMotov** [comment, marked as answer]
You should first define all operations for your router and only then include it into app.

So, you just need to move `app.include_router(router)` down, so that it would be below the endpoint definition:

```python
from typing import Annotated

from fastapi import FastAPI, APIRouter, Depends
from fastapi.testclient import TestClient
from fastapi.security import OAuth2PasswordRequestForm


app = FastAPI()
router = APIRouter()


@router.post("/token")  # working
def login_for_access_token(
    form_data: Annotated[OAuth2PasswordRequestForm, Depends()],
):
    return form_data


app.include_router(router)  # Include router after defining its routes


client = TestClient(app)


def test_token_success():
    response = client.post(
        "/token",
        data={"username": "johndoe", "password": "secret"}
    )
    print(f"response: {response.json()}")
    assert response.status_code == 200

```
